ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/commons/RaoUtil.java

```diff
@@ -128,6 +125,39 @@ public static boolean isAnyMarginNegative(FlowResult flowResult, Set<FlowCnec> f
         return flowCnecs.stream().anyMatch(flowCnec -> flowResult.getMargin(flowCnec, marginUnit) <= 0);
     }
 
+    /**
+     * Evaluates if a remedial action is available.
+     * 1) If the remedial action has no usage rule, it will not be available.
+     * 2) Then we separate usageRules in two categories :
+     *      - OnInstant and OnContingencyState (OnState category)
+     *      - OnFlowConstraint and OnFlowConstraintInCountry (OnFlow category)
+     * For automatonState, we do not want to check the OnState category as they are supposed to be "FORCED"
+     * For the OnFlow category, we want to check their availability for every given state.
+     * Therefore, we perform a double check : The state must not be an automatonState.
+     * If so, it gathers every OnState usage rules and
+     * all it takes is for one to be "AVAILABLE" for the remedial action to be available.
+     * 3) Else, it calls the method RemedialAction.isRemedialActionAvailable to give the verdict.
+     */
+    public static boolean isRemedialActionAvailable(RemedialAction<?> remedialAction, State state, PrePerimeterResult prePerimeterResult, Set<FlowCnec> flowCnecs, Network network, RaoParameters raoParameters) {
+        List<UsageRule> usageRules = remedialAction.getUsageRules();
+        if (usageRules.isEmpty()) {
+            FaraoLoggerProvider.BUSINESS_WARNS.warn(String.format("The remedial action %s has no usage rule and therefore will not be available.", remedialAction.getName()));
+            return false;
+        }
+        Set<UsageRule> usageRulesOnTime = usageRules.stream()
+            .filter(usageRule -> usageRule instanceof OnContingencyState || usageRule instanceof OnInstant).collect(Collectors.toSet());
+        if (state.getInstant().equals(Instant.AUTO)) {
+            if (usageRulesOnTime.stream().anyMatch(usageRule -> usageRule.getUsageMethod(state).equals(UsageMethod.FORCED))) {
+                return true;
+            }
+        } else if (usageRulesOnTime.stream()
+            .anyMatch(usageRule -> usageRule.getUsageMethod(state).equals(UsageMethod.AVAILABLE)
+                || usageRule.getUsageMethod(state).equals(UsageMethod.FORCED))) {
+            return true;
+        }
+        return remedialAction.isRemedialActionAvailable(state, isAnyMarginNegative(prePerimeterResult, remedialAction.getFlowCnecsConstrainingUsageRules(flowCnecs, network, state), raoParameters.getObjectiveFunctionParameters().getType().getUnit()));
```

Comment:
I think there's a case not covered by this logic : if a RA has an OnInstant AVAILABLE usage rule, and another OnFlowConstraint UNAVAILABLE usage rule. If the CNEC is constrained, the RA should not be available.